sherlock-webhook-proxy/pkg/handler.go

```diff
@@ -0,0 +1,248 @@
+				log.Printf("library said hook was sent with an invalid method\n")
+				return
+			default:
+				w.WriteHeader(http.StatusInternalServerError)
+				log.Printf("library had an unknown error handling the hook: %v\n", err)
+				return
+			}
+		}
+
+		// Check HMAC + SHA256 signature
+		calculatedSignature := hex.EncodeToString(_mac.Sum(nil))
+		if signature != calculatedSignature {
+			w.WriteHeader(http.StatusForbidden)
+			log.Printf("HMAC + SHA 256 signature was %s but calculated was %s, rejecting\n", signature, calculatedSignature)
+			return
+		}
+
+		// Handle parsed payloads from the library
+		switch payload := rawPayload.(type) {
+
+		// ping issued upon the webhook being added to a new repo; might as well respond with 200
+		case github.PingPayload:
+			w.WriteHeader(http.StatusOK)
+			log.Printf("received ping from repo %s", payload.Repository.FullName)
+
+		// workflow_run issued upon workflow request, running, and completion
+		case github.WorkflowRunPayload:
+			if token, present := os.LookupEnv(iapTokenOverrideEnvVar); present {
+				// If we have a token, just use that
+				_transport.DefaultAuthentication = httptransport.BearerToken(token)
+			} else {
+				// Otherwise, do the dance to get it from the metadata server
+				formedIdTokenUrl := fmt.Sprintf("%s?audience=%s", idTokenUrl, os.Getenv(iapAudienceEnvVar))
+				req, err := http.NewRequest(http.MethodGet, formedIdTokenUrl, nil)
+				if err != nil {
+					w.WriteHeader(http.StatusInternalServerError)
+					log.Printf("http.NewRequest(%s, %s): %v\n", http.MethodGet, formedIdTokenUrl, err)
+					return
+				}
+				req.Header.Set("Metadata-Flavor", "Google")
+				resp, err := (&http.Client{}).Do(req)
+				if err != nil {
+					w.WriteHeader(http.StatusInternalServerError)
+					log.Printf("(&http.Client{}).Do(%s): %v\n", formedIdTokenUrl, err)
+					return
+				} else if resp.StatusCode != http.StatusOK {
+					w.WriteHeader(http.StatusInternalServerError)
+					log.Printf("(&http.Client{}).Do(%s): non-200: %d", formedIdTokenUrl, resp.StatusCode)
+					return
+				}
+				idToken, err := io.ReadAll(resp.Body)
+				_ = resp.Body.Close()
+				if err != nil {
+					w.WriteHeader(http.StatusInternalServerError)
+					log.Printf("io.ReadAll(resp.Body): %v\n", err)
+					return
+				}
+				_transport.DefaultAuthentication = httptransport.BearerToken(string(idToken[:]))
+			}
+
+			sherlockClient := client.New(_transport, strfmt.Default)
+
+			// Convert webhook fields into what we'll store in Sherlock
+			var status, terminalAt string
+			if payload.WorkflowRun.Conclusion != "" {
+				status = payload.WorkflowRun.Conclusion
+				terminalAt = payload.WorkflowRun.UpdatedAt.Format(time.RFC3339)
+			} else {
+				status = payload.WorkflowRun.Status
+			}
+
+			// PUT to Sherlock; will create if the selector isn't found and edit otherwise
+			edited, created, err := sherlockClient.CiRuns.PutAPIV2CiRunsSelector(&ci_runs.PutAPIV2CiRunsSelectorParams{
```

Comment:
if there's ultimately still a 1:1 ratio between GitHub requests and requests to sherlock, I guess the advantage of having this in a cloud function is that:

(1) we reject invalid/unauthed requests before they even reach Sherlock, which protects from abuse
(2) we don't have to poke a hole in IAP, which protects us from accidentally making that hole bigger than it should be

is that correct?

Reply:
Yeah, my whole idea for serverless here is that it makes this a lower risk change, because it doesn't actually change Sherlock's security model. Sherlock's IAP isn't any less effective at protecting it after this change. My other idea is that it's super easy to modify this cloud function to handle other payloads, that potentially don't have that 1:1 relationship (for example, could subscribe to `issue_comment` payloads and increment a per-repo counter in sherlock whenever `jenkins retest` is commented)